src/discussions/comments/data/slices.js

```diff
@@ -3,14 +3,28 @@ import { createSlice } from '@reduxjs/toolkit';
 import { RequestStatus } from '../../../data/constants';
 
 function normaliseComments(state, rawCommentData) {
-  const { threadCommentMap: threads, comments } = state;
+  const {
+    threadCommentMap: threads,
+    commentResponsesMap: responses,
+    comments,
+  } = state;
   rawCommentData.forEach(
     comment => {
-      if (!threads[comment.thread_id]) {
-        threads[comment.thread_id] = [];
-      }
-      if (!threads[comment.thread_id].includes(comment.id)) {
-        threads[comment.thread_id].push(comment.id);
+      if (comment.parent_id) {
+        if (!responses[comment.parent_id]) {
+          responses[comment.parent_id] = [];
+        }
+        if (!responses[comment.parent_id].includes(comment.id)) {
+          responses[comment.parent_id].push(comment.id);
+        }
+      } else {
+        if (!threads[comment.thread_id]) {
+          threads[comment.thread_id] = [];
+        }
+
+        if (!threads[comment.thread_id].includes(comment.id)) {
+          threads[comment.thread_id].push(comment.id);
+        }
       }
       comments[comment.id] = comment;
     },
```

Comment:
This code is reformatting the API response into something that's easier for the UI to look up. 

So, when we call an API and get a list of comments, here is what we do:
- We add it to the comment map using its id. So if we ever need to look up the comment by its id, we can just do state.comments[commentId] to get the comment.
- If the comment has a parent id, it is a comment on a comment, i.e. a response. So, we add it to the list of replies to a comment. For this, we maintain a dictionary of comment ids, to the responses to those comments. 
- If the comment doesn't have a parent id it is a comment on the main thread. So, we add it to a list of the comments in the that thread as above. 

So we take the following three comments:
- {id: "comment1", thread_id: "tid1", parent_id: null, ...}
- {id: "response1", thread_id: "tid1", parent_id: "comment1", ...}
- {id: "comment2", thread_id: "tid1", parent_id: null, ...}

And create the following:
```javascript 
    comments = { 
        comment1: {id: "comment1", thread_id: "tid1", parent_id: null, ...},
        response1: {id: "response1", thread_id: "tid1", parent_id: "comment1", ...}
        comment2: {id: "comment2", thread_id: "tid1", parent_id: null, ...}
    }
    threadCommentMap = {
        tid1: ["comment1", "comment2"]  //note doesn't list response1
    }
    commentResponseMap = {
        comment1: ["response1"]
    }
```

Reply:
Can this be added as a comment or documentation?

I feel like encapsulating the map lookups would make this much easier to read and use.
For example, the `selectCommentResponses` and `selectThreadComments` selectors would become obsolete because you could instead use
```js
const selectComments = state => state.comments;

// then, in component...
const comments = useSelector(selectComments);
const responses = comments.getResponsesTo(commentId); // returns [] if there are no responses
```

Reply:
I don't think redux state can handle that. I do not think you can have arbitrary functions be part of redux state. Redux state should be plain javascript objects so that redux can reason about their state and deteming if/when the state has changed to propagate ui update events. 

Of course, we can encapsulate all the code to deal with the redux in a class, but I'm not sure how that helps. React/redux are generally more functional in paradigm than object-oriented and it'll be an ill fit.

Reply:
Maybe it can be embedded into the selector, then?

Reply:
It is right? We ahev `selectCommentResponses` that takes a `commentId`, so you can use it as `useSelector(selectCommentResponses(commentId))`.

Reply:
You're right!